SYCL/Plugin/interop-experimental-single-TU-SYCL-CUDA-compilation.cpp

```diff
@@ -0,0 +1,207 @@
+}
+
+__global__ void test_cuda_kernel(int *out) {
+  int i = blockIdx.x * blockDim.x + threadIdx.x;
+  out[i] = i - test_cuda_function_1();
+}
+
+// Test CUDA kernel launch and CUDA API.
+void test_cuda_kernel_launch() {
+  // CUDA
+  const int n = 512;
+  std::vector<int> result(n, -1);
+  int *cuda_kern_result = NULL;
+
+  int block_size = 128;
+  dim3 dimBlock(block_size, 1, 1);
+  dim3 dimGrid(n / block_size, 1, 1);
+
+  cudaMalloc((void **)&cuda_kern_result, n * sizeof(int));
+
+  test_cuda_kernel<<<n / block_size, block_size>>>(cuda_kern_result);
+
+  cudaError_t error = cudaGetLastError();
+  if (error != cudaSuccess)
+    std::cerr << "CUDA ERROR: " << error << " " << cudaGetErrorString(error)
+              << std::endl;
+
+  cudaMemcpy(result.data(), cuda_kern_result, n * sizeof(int),
+             cudaMemcpyDeviceToHost);
+
+  for (size_t i = 0; i < n; i++)
+    assert((0 == result[i]) && "Kernel execution fail!");
+
+  cudaFree(cuda_kern_result);
+}
+
+// ------------------------------------------------------------------------- //
+
+__host__ float test_cuda_function_7() { return -1; }
+__device__ float test_cuda_function_7() { return 1; }
+
+__host__ float test_cuda_function_8() { return 3; }
+
+__device__ float test_cuda_function_9() { return 9; }
+
+int test_regular_function_0() { return test_cuda_function_7(); }
+
+int test_regular_function_1() { return test_cuda_function_8(); }
+
+int test_regular_function_2() { return test_cuda_function_9(); }
+
+// Test the correct emission of __device__/__host__ function when called by
+// regular functions.
+void test_regular_functions(sycl::queue &q) {
+
+  // regular func must returning the __host__ one (so, 1.0f)
+  assert((test_regular_function_0() == -1) &&
+         "Mismatch regular func to __host__");
+  assert((test_regular_function_1() == 3) &&
+         "Mismatch regular func to __host__");
+
+  sycl::buffer<int, 1> b_r{3};
+  q.submit([&](sycl::handler &h) {
+    sycl::accessor r{b_r, h, sycl::write_only};
+
+    h.single_task([=]() {
+      r[0] = test_regular_function_1(); //<-- points to __host__
+      r[1] = test_regular_function_2(); //<-- points to __device__
+      r[2] = test_regular_function_0(); //<-- points to __device__
+    });
+  });
+
+  sycl::host_accessor r{b_r, sycl::read_only};
+  assert((r[0] == 3) && "Mismatch regular func to __host__");
+  assert((r[1] == 9) && "Mismatch regular func to __device__");
+  assert((r[2] == 1) && "Mismatch regular func to __device__");
+}
+
+// ------------------------------------------------------------------------- //
+
+// Tests the result of a function that calls CUDA device builtins.
+void test_ids(sycl::queue &q) {
+
+  const size_t n1 = 2048;
+  const sycl::range<1> r1{n1};
+  sycl::buffer<int, 1> b_idx{n1};
+  q.submit([&](sycl::handler &h) {
+    sycl::accessor d_idx{b_idx, h, sycl::write_only};
+
+    h.parallel_for(r1,
+                   [=](sycl::id<1> i) { d_idx[i] = test_cuda_function_1(); });
```

Comment:
Where is the dimensions mapping specified?

Reply:
Since this is a one dimensional range it maps to the x-dimension in CUDA.

Further details, which are valid also for `dpc++` can be found [here](https://codeplay.com/portal/blogs/2019/11/18/computecpp-v1-1-6-changes-to-work-item-mapping-optimization.html).